sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/BaseExplodedGraph.cs

```diff
@@ -1,442 +1,442 @@
+                ProgramState = programState
+            });
+        }
+
+        protected void OnInstructionProcessed(SyntaxNode instruction, ProgramPoint programPoint, ProgramState programState)
+        {
+            InstructionProcessed?.Invoke(this, new InstructionProcessedEventArgs
+            {
+                Instruction = instruction,
+                ProgramPoint = programPoint,
+                ProgramState = programState
+            });
+        }
+
+        protected void OnConditionEvaluated(SyntaxNode condition, bool evaluationValue)
+        {
+            ConditionEvaluated?.Invoke(this, new ConditionEvaluatedEventArgs
+            {
+                Condition = condition,
+                EvaluationValue = evaluationValue
+            });
+        }
+
+        #endregion
+
+        #region Visit*
+
+        protected abstract void VisitBinaryBranch(BinaryBranchBlock binaryBranchBlock, ExplodedGraphNode node);
+
+        protected abstract void VisitInstruction(ExplodedGraphNode node);
+
+        protected virtual void VisitBranchBlock(ExplodedGraphNode node, ProgramPoint programPoint)
+        {
+            var newProgramState = node.ProgramState.PopValue();
+            newProgramState = CleanStateAfterBlock(newProgramState, node.ProgramPoint.Block);
+            EnqueueAllSuccessors(programPoint.Block, newProgramState);
+        }
+
+        protected virtual void VisitBranchBlock(BranchBlock branchBlock, ExplodedGraphNode node)
+        {
+            var newProgramState = node.ProgramState;
+            if (IsValueConsumingStatement(branchBlock.BranchingNode))
+            {
+                newProgramState = newProgramState.PopValue();
+            }
+            newProgramState = CleanStateAfterBlock(newProgramState, branchBlock);
+            EnqueueAllSuccessors(branchBlock, newProgramState);
+        }
+
+        protected virtual void VisitSingleSuccessorBinaryBranch(BinaryBranchingSimpleBlock block, ExplodedGraphNode node)
+        {
+            SymbolicValue sv;
+            var programState = node.ProgramState.PopValue(out sv);
+
+            foreach (var newProgramState in sv.TrySetConstraint(BoolConstraint.True, programState))
+            {
+                OnConditionEvaluated(block.BranchingInstruction, evaluationValue: true);
+                var nps = newProgramState.PushValue(SymbolicValue.True);
+                EnqueueNewNode(new ProgramPoint(block.SuccessorBlock), nps);
+            }
+
+            foreach (var newProgramState in sv.TrySetConstraint(BoolConstraint.False, programState))
+            {
+                OnConditionEvaluated(block.BranchingInstruction, evaluationValue: false);
+                var nps = newProgramState.PushValue(SymbolicValue.False);
+                EnqueueNewNode(new ProgramPoint(block.SuccessorBlock), nps);
+            }
+        }
+
+        protected virtual void VisitSimpleBlock(SimpleBlock block, ExplodedGraphNode node)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'SonarAnalyzer.Helpers.FlowAnalysis.Common.Block' instead of 'SonarAnalyzer.Helpers.FlowAnalysis.Common.SimpleBlock'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)